functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,167 @@
+package transformer
+
+import (
+	"fmt"
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+// the similar struct esxits in resid.GVK, but there is no function to create an GVK struct without using kyaml
+type FieldSpec struct {
+	Group              string `json:"group,omitempty" yaml:"group,omitempty"`
+	Version            string `json:"version,omitempty" yaml:"version,omitempty"`
+	Kind               string `json:"kind,omitempty" yaml:"kind,omitempty"`
+	Path               string `json:"path,omitempty" yaml:"path,omitempty"` // seperated by /
+	CreateIfNotPresent bool   `json:"create,omitempty" yaml:"create,omitempty"`
+}
+
+func SetLabels(rl *fn.ResourceList) (bool, error) {
+	transformer := LabelTransformer{}
+	if err := transformer.Config(rl.FunctionConfig); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, nil
+	}
+	if err := transformer.Transform(rl.Items); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, nil
+	}
+	return true, nil
+}
+
+type LabelTransformer struct {
+	// Desired labels
+	NewLabels map[string]string `json:"labels,omitempty" yaml:"labels,omitempty"`
+	// FieldSpecs is deprecated, please use AdditionalLabelFields instead.
+	FieldSpecs []FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+	// AdditionalLabelFields is used to specify additional fields to add labels.
+	AdditionalLabelFields []FieldSpec `json:"additionalLabelFields,omitempty" yaml:"additionalLabelFields,omitempty"`
+	// Results is used to track labels that have been applied
+	Results fn.Results
+}
+
+// Config parse the functionConfig kubeObject to the fields in the LabelTransformer
+func (p *LabelTransformer) Config(o *fn.KubeObject) error {
+	// parse labels to NewLabels
+	switch {
+	case o.IsEmpty():
+		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap` or `SetNamespace`")
+	case o.IsGVK("", "", "ConfigMap"):
+		p.NewLabels = o.NestedStringMapOrDie("data")
```

Comment:
I have tried using o.As(&cm). But it is not working because the `corev1.ConfigMap` is not has yaml annotation. So the As() failed. 
the comfigMap struct's Data field is: `Data map[string]string `json:"data,omitempty" protobuf:"bytes,2,rep,name=data"``, it doesn't has yaml. 
For the runner, I have looked into the code in details. Their functionality is pretty similar, I think they have the same level of simplicity. 
Do I need to switch the implementation to Runner? it is about changing 20-30 lines of code.

Reply:
😄 Yeah, this is one of the reasons I mentioned before about getting rid of the `yaml.Unmarshal`, so we do not have this limitations. Two changes are needed before dropping yaml, one is disabling the `AdditionalFieldSpec`, which Mengqi has verified it's safe to do so, the other is changing the const `commonLabelFieldSpecs`, you can change that to the `set-namespace` approach by comparing the GVK, and nestedFields.     

Using `fn.Runner` should much simplify the logic and give a good example for other fn authors.